Redesign of email settings

raphaelm opened the pull request:
This tries to "solve email":tm:, i.e. make all the common misconfigurations and problems go away. This requires a redesign of how the settings around email work because it's not really possible to view "sender address" and "SMTP server" as distinct concepts, validation needs to look at them in combination.

There is one regression in this PR, it is now no longer allowed to set mail_from through the API, but I see no way around this for the moment.

Main changes:

- Allows sysadmins to set different sender address depending on whether an email is related to the system, an event or organizer, or a user notification
  - This will allow us to better deal with autoresponders replying to every notification
  - This will allow us to use a separate sender address for user-generated mail
- Incorrect or missing SPF records now show a helpful warning (if turned on by sysadmin)
- Ownership of sender addresses is validated before they can be used (can be turned off by sysadmin)
  - It is no longer possible by default to use a private IP address as an SMTP host since that would allow to circumvent validation (can be turned off by sysadmin)
- It is no longer possible to set a non-working SMTP server as active
- A warning is shown if GMail is used as SMTP (GMail's rate limits are a common source of issues)

There's no change to the underlying settings data structures, just to the user interface. Therefore no migration is needed.

Here are some impressions:

## Settings page

![Screenshot 2022-01-22 at 15-39-49 Organizer pretix de](https://user-images.githubusercontent.com/64280/150648451-14dc07a6-c72d-49a5-957a-b77a78de1284.png)

## System default

![Screenshot 2022-01-22 at 17-21-11 Demo pretix de](https://user-images.githubusercontent.com/64280/150648478-8ce78735-9f2d-4696-bfdb-8e968437e9e6.png)

## Custom sender, no SMTP

![Screenshot 2022-01-22 at 17-21-28 Demo pretix de](https://user-images.githubusercontent.com/64280/150648489-29cddd89-6ee9-4b08-a043-5d76b3de3004.png)
![Screenshot 2022-01-22 at 17-59-53 Organizer pretix de](https://user-images.githubusercontent.com/64280/150648499-b31fb8bd-3383-4700-ae7e-4c214fda8c43.png)
![Screenshot 2022-01-22 at 17-59-04 Organizer pretix de](https://user-images.githubusercontent.com/64280/150648500-4bcbcac1-7776-4724-9ebc-442ab88c8b8d.png)
![Screenshot 2022-01-22 at 17-57-38 Organizer pretix de](https://user-images.githubusercontent.com/64280/150648501-10d90ca5-41f2-4764-bb6e-944e4896c194.png)


## Custom SMTP

![Screenshot 2022-01-22 at 16-59-19  pretix de](https://user-images.githubusercontent.com/64280/150648454-83e70192-5a52-4c8a-ba36-70aa676e5413.png)
![Screenshot 2022-01-22 at 16-59-51  pretix de](https://user-images.githubusercontent.com/64280/150648458-f1cfdbfc-27f5-49e6-bf6b-11d4f9459a5e.png)
![Screenshot 2022-01-22 at 17-17-18  pretix de](https://user-images.githubusercontent.com/64280/150648467-686cc9bf-609f-470b-84d5-6d2e8826e34d.png)
![Screenshot 2022-01-22 at 17-00-13  pretix de](https://user-images.githubusercontent.com/64280/150648470-7b2e6908-eb66-400e-ac6f-ab85e36b6408.png)

pc-coholic commented:
Without having done any testing yet and just having skimmed through the code: should we perhaps make the SPF-check a little bit more comprehensive by actually checking HELO and IP instead of presence of a certain string?

We have seen customers that have so many external services that send on their behalf, that we are not present on the "main" SPF-record, but included somewhere down the line/DNS-lookups...

Reference: pretixeu/support/tasks.py:spf_checker()

felixrindt commented on src/pretix/control/views/mailsetup.py:
```diff
@@ -0,0 +1,219 @@
+                            'smtp.gmail.com': _(
+                                'We recommend not using Google Mail for transactional emails. If you try sending many '
+                                'emails in a short amount of time, e.g. when sending information to all your ticket '
+                                'buyers, there is a high change Google will not deliver all of your emails since they '
```
```suggestion
                                'buyers, there is a high chance Google will not deliver all of your emails since they '
```

raphaelm commented:
> Without having done any testing yet and just having skimmed through the code: should we perhaps make the SPF-check a little bit more comprehensive by actually checking HELO and IP instead of presence of a certain string?

I'm actually not so sure on this one. I wasn't **just** trying to avoid an additional dependency, I also had some motive for not doing a "proper" SPF check. The problem is that with a proper SPF checker like PySPF we'd be validating a different thing than we actually want to validate.

Let's say that one of our outgoing pretix.eu SMTP servers is 12.34.56.78. If a client now includes ``ip4:12.34.56.78`` in their SPF record, they would pass the validation if we check with PySPF. However, I want them to fail. I only want them to pass if they literally included ``include:_spf.pretix.eu`` and not anything else, because otherwise they will run into problems if we ever add IP addresses.

So presence of a certain string is actually what I was going for here. The only caveat is that we won't detect the string if it's hidden behind another layour of ``include:``. We could solve this by implementing ``include:`` ourselves, or by just accepting it since it's a dismissable warning and tech-savvy people who have long SPF records might be fine ignoring it.

I'm also not sure if the HELO case is relevant, because the HELO domain will always be "our" domain and it's not up to the customer whether that has a correct SPF record. AFAIK our HELO domains do not have an SPF record right now (I did not know this was a thing!) but that did not hurt so far.

Also:

> We have seen customers that have so many external services that send on their behalf, that we are not present on the "main" SPF-record, but included somewhere down the line/DNS-lookups...

Really? I don't remember *this* kind of case, maybe we can talk about that tomorrow. I also don't think there's a really ever a problem adding us to the "main" record. Some sources on the internet say the record may not be more than 255 chars, but that is wrong. While in fact a "DNS string" may only have 255 characters, a DNS record may be composed of multiple strings: https://datatracker.ietf.org/doc/html/rfc7208#section-3.3

RFC7208 *recommends* keeping it below 450-512 bytes, but actually our DNS record at pretix.eu has been 678 bytes for a while and we haven't encountered a single problem so that might not be really relevant any more. https://datatracker.ietf.org/doc/html/rfc7208#section-3.4

pc-coholic reviewed (commented):
- When a custom sender-address has been set on the organizer-level, the event-level overview is not indicating that the current setting has been inherited. Probably a question of cosmetics, if we want to expose that here.
- However: if a custom-sender address has been set on the organizer-level and I then proceed to change it on the event-level, "Use system default" will be selected on the next screen and not (as expected by me) the "Use system email server with a custom sender address"
- I think the event-level setting should have a fourth option "Follow organizer-level settings", which effectively clears the event-level settings. As of right now, I have no way to revert event-level settings into a state, where the organizer-level settings are used once again.

raphaelm commented on src/pretix/control/templates/pretixcontrol/email/email_setup.txt:
```diff
@@ -0,0 +1,14 @@
+{% load i18n %}{% blocktrans with code=code instance=instance %}Hello,
+
+someone requested to use this email address as a sender address on {{ instance }}.
```
good point

raphaelm commented:
>    When a custom sender-address has been set on the organizer-level, the event-level overview is not indicating that the current setting has been inherited. Probably a question of cosmetics, if we want to expose that here.

Yes it does:
![2022-01-25-205306_1351x953_scrot](https://user-images.githubusercontent.com/64280/151049412-e764d3b5-c951-4ef1-bf0e-d1aa1a83120b.png)

I've now split it into two blocks which looks a little weird but is more useful

>    However: if a custom-sender address has been set on the organizer-level and I then proceed to change it on the event-level, "Use system default" will be selected on the next screen and not (as expected by me) the "Use system email server with a custom sender address"

Fixed

>    I think the event-level setting should have a fourth option "Follow organizer-level settings", which effectively clears the event-level settings. As of right now, I have no way to revert event-level settings into a state, where the organizer-level settings are used once again.

Well, that's the case for all settings in pretix, not specific to these. Still easy enough, added it in.
